packages/cli/src/cog/cog.job.ts

```diff
@@ -0,0 +1,340 @@
+import {
+    FileConfig,
+    FileConfigPath,
+    FileOperator,
+    ProjectionTileMatrixSet,
+    titleizeImageryName,
+} from '@basemaps/shared';
+import { Projection } from '@basemaps/shared/build/proj/projection';
+import { MultiPolygon, multiPolygonToWgs84, Wgs84 } from '@linzjs/geojson';
+import { CliInfo } from '../cli/base.cli';
+import { GdalCogBuilderDefaults, GdalCogBuilderResampling } from '../gdal/gdal.config';
+import {
+    CogBuilderMetadata,
+    CogGenerated,
+    CogOutputProperties,
+    CogProperties,
+    CogSourceProperties,
+    CogStac,
+    CogStacExtensions,
+    CogStacKeywords,
+    CogStacLicense,
+    CogStacVersion,
+    StacLink,
+    StacProvider,
+    StacCollection,
+    CogBuild,
+} from './types';
+
+export const MaxConcurrencyDefault = 50;
+
+export interface JobCreationContext {
+    /** Source config */
+    sourceLocation: FileConfig | FileConfigPath;
+
+    /** Output config */
+    outputLocation: FileConfig;
+
+    /** Should the imagery be cut to a cutline */
+    cutline?: {
+        href: string;
+        blend: number;
+    };
+
+    targetProjection: ProjectionTileMatrixSet;
+
+    override?: {
+        /** Override job id */
+        id?: string;
+
+        /**
+         * Image quality
+         * @default GdalCogBuilderDefaults.quality
+         */
+        quality?: number;
+
+        /**
+         * Number of threads to use for fetches
+         * @default MaxConcurrencyDefault
+         */
+        concurrency?: number;
+
+        /**
+         * Override the source projection
+         */
+        projection?: Epsg;
+
+        /**
+         * Resampling method
+         * @Default  GdalCogBuilderDefaults.resampling
+         */
+        resampling?: GdalCogBuilderResampling;
+    };
+
+    /**
+     * Should this job be submitted to batch now?
+     * @default false
+     */
+    batch: boolean;
+
+    /** Should this job ignore source coverage and just produce one big COG for EPSG extent */
+    oneCog: boolean;
+}
+
+/**
+ * Attempt to parse a year from a imagery name
+ * @example wellington_urban_2017_0.10m -> 2017
+ * @param name Imagery name to parse
+ * @return imagery year, -1 for failure to parse
+ */
+export function extractYearRangeFromName(name: string): [number, number] {
+    const re = /(?:^|\D)(\d{4})(?:-(\d{2}))?(?:$|\D)/g;
+
+    const years: number[] = [];
+
+    for (let m = re.exec(name); m != null; m = re.exec(name)) {
+        years.push(parseInt(m[1]));
+        if (m[2] != null) years.push(parseInt(m[1].slice(0, 2) + m[2]));
+    }
+
+    if (years.length == 0) {
+        return [-1, -1];
+    }
+
+    years.sort();
+
+    return [years[0], years[years.length - 1] + 1];
+}
+
+export class CogJob {
+    stac: CogStac;
+    private cacheTargetZoom?: {
+        gsd: number;
+        zoom: number;
+    };
+
+    /**
+     * The Name to used when just producing one cog to cover the full extent
+     */
+    static OneCogName = '0-0-0';
+
+    static async build({
```

Comment:
not sold on this syntax, could you break it out into a type and add some comments about what the things are for